### PR TITLE
#182461136 - USER SHOULD BE ABLE TO REGISTER SUCCESSFULLY

### PR DESCRIPTION
**Pull request by dalienst**
#### What does this Pull Request do?

- [ ] User registration and log in

#### Description of Task to be completed?

- [ ] User is able to register successfully and receive a token upon registration.

#### How should this be manually tested?

- [ ] Testing endpoints via Postman

#### Any background context you want to provide?

- [ ] Registration of users

#### What are the relevant pivotal tracker?
[#182461136](https://www.pivotaltracker.com/n/projects/2570600/stories/182461136)
#### Screenshots (if appropriate)

- [ ] N/A

#### Questions:

- [ ] N/A

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -12,6 +12,7 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+djangorestframework-simplejwt = "*"
```
Comment (mainamuriuki274) [inline Pipfile:15]: version?

File: app/user/models.py [hunk 2]
```diff
@@ -42,6 +42,7 @@ def create_superuser(
     ) -> Any:
         kwargs.setdefault("is_staff", True)
         kwargs.setdefault("is_superuser", True)
+        kwargs.setdefault("is_active", True)
```
Comment (mainamuriuki274) [inline app/user/models.py:45]: No need for this if default is True

File: app/user/serializers.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from django.contrib.auth import get_user_model
+from django.core.mail import send_mail
+from django.template.loader import render_to_string
+from rest_framework import serializers
+from rest_framework.validators import UniqueValidator
+
+from speaksfer.settings.base import EMAIL_USER
+
+User = get_user_model()
+Profile = get_user_model()
```
Comment (mainamuriuki274) [inline app/user/serializers.py:12]: No only the user model is gotten this way, other models are gottebn the normal way using relative imports

File: app/user/serializers.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        max_length=128,
+        min_length=8,
+        write_only=True,
+    )
+
+    class Meta:
+        model = User
+        fields = ("email", "username", "password")
+
+    # def validate(self, data):
```
Comment (mainamuriuki274) [inline app/user/serializers.py:37]: remove unused comments

File: app/user/serializers.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        user = User.objects.create_user(**validated_data)
+        self.send_email(user)
+
+        return user
+
+
+class ProfileSerializer(serializers.ModelSerializer):
+    username = serializers.CharField(source="user.username")
+    bio = serializers.CharField(allow_blank=True, required=False)
+    image = serializers.SerializerMethodField()
```
Comment (mainamuriuki274) [inline app/user/serializers.py:66]: where is the method?

File: app/user/templates/email_verification.html [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{% autoescape off %}
+
+Hello {{ user.username }},
+
+Verify your email  by clicking the link below:
+{% block reset_link %}
+https://speaksfer/hi
```
Comment (mainamuriuki274) [inline app/user/templates/email_verification.html:7]: enter a valid url that has the token

File: app/user/views.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        response["refresh"] = str(refresh)
+        response["access"] = str(refresh.access_token)
+
+        return Response(response, status=status.HTTP_201_CREATED)
+
+
+class VerifyEmail(APIView):
+    def get(self, request: Request) -> Response:
+        token = request.GET.get("token")
+        try:
```
Comment (mainamuriuki274) [inline app/user/views.py:31]: You can have this in the serializers.py in VerifyEmailSerialier

File: speaksfer/settings/base.py [hunk 8]
```diff
@@ -41,14 +42,14 @@
     "django.contrib.sessions",
     "django.contrib.messages",
     "django.contrib.staticfiles",
-    # third-party imports
-    "drf_yasg",
+    "app.user",
```
Comment (mainamuriuki274) [inline speaksfer/settings/base.py:45]: why did you remove the comment?

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/user/templates/email_verification.html (hunk 6 above)
Comment (dalienst) [reply app/user/templates/email_verification.html:7]: How do I do that?

File: speaksfer/settings/base.py (hunk 8 above)
Comment (dalienst) [reply speaksfer/settings/base.py:45]: It was conflicting with the name of the app but let me fix it

File: Pipfile [hunk 9]
```diff
@@ -12,6 +12,8 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+djangorestframework-simplejwt = "5.2.0"
+types-six = "*"
```
Comment (mainamuriuki274) [inline Pipfile:16]: remove types-six use smart bytes/ smart str django in built functions

File: app/user/serializers.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+from app.user.models import Profile
+from app.user.token import account_activation_token
+from speaksfer.settings.base import EMAIL_USER
+
+User = get_user_model()
+
+
+class UserSerializer(serializers.ModelSerializer):
+
+    username = serializers.CharField(
```
Comment (mainamuriuki274) [inline app/user/serializers.py:22]: Shouldn't the username be unique?

File: app/user/serializers.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            email_body,
+            EMAIL_USER,
+            [user.email],
+            fail_silently=False,
+        )
+
+    def create(self, validated_data: Any) -> Any:
+        request = self.context.get("request")
+        user = User.objects.create_user(**validated_data)
+        self.send_email(user, request)  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/serializers.py:67]: Do you really neeed to ignore, also do not just ignore all ignore a specific code

File: app/user/serializers.py (hunk 5 above)
Comment (mainamuriuki274) [reply app/user/serializers.py:75]: Then why do you have the ProfileSerializer?

File: app/user/serializers.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        max_length=20,
+        min_length=8,
+    )
+
+    email = serializers.EmailField(
+        required=True,
+        validators=[UniqueValidator(queryset=User.objects.all())],
+    )
+
+    password = serializers.CharField(
```
Comment (mainamuriuki274) [inline app/user/serializers.py:32]: Include other password validators, check if it has a number, uppercase, lowercase, symbol

File: app/user/serializers.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        except (TypeError, ValueError, OverflowError, User.DoesNotExist):
+            raise serializers.ValidationError(
+                "Invalid user id", code="invalid_code"
+            )
+
+        token = data.get("token")
+
+        if user and account_activation_token.check_token(user, token):
+            user.is_verified = True
+            user.save()
```
Comment (mainamuriuki274) [inline app/user/serializers.py:105]: Override the save functionality for this, only validate token and uidb64 are valid here

File: app/user/tests/test_serializer.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
+from django.contrib.auth import get_user_model
+from django.test import TestCase
+
+from app.user.models import Profile
+
+User = get_user_model()
+
+
+class ImageUploadTest(TestCase):
+    def test_get_image(self):  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/tests/test_serializer.py:10]: no need for ignore everywhere it raises an errror otherwise what's the point of using mypy

File: app/user/tests/test_serializer.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        user = User.objects.create(
+            username="bearhands",
```
Comment (mainamuriuki274) [inline app/user/tests/test_serializer.py:12]: use faker library to generate fake data

File: app/user/tests/test_serializer.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+            email="bear@gmail.com",
+            password="bearpassword",
+        )
+        data = Profile.objects.create(
+            image="images/sample.jpg", bio="Hello mellow", user_id=user.id
```
Comment (mainamuriuki274) [inline app/user/tests/test_serializer.py:17]: use faker library

File: app/user/tests/test_serializer.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+        )
+        response = self.client.get(data.image)
```
Comment (mainamuriuki274) [inline app/user/tests/test_serializer.py:19]: if this is for the serializer is there are need for the client, check out when to use Client class

File: app/user/tests/test_views.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    def setUp(self) -> None:
+        self.test_user = User.objects.create_user(
+            "testuser", "test@mail.com", "tester012"
+        )
+        self.create_url = reverse("register")
+
+    # Testing UserView
+    def test_create_user(self) -> None:
+        data = {
+            "username": "foobarata",
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:25]: use faker library

File: app/user/tests/test_views.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+            "email": "foobar@example.com",
+            "password": "somepassword",
+        }
+
+        response = self.client.post(self.create_url, data, format="json")
+
+        self.assertNotEqual(response, status.HTTP_201_CREATED)
+        self.assertEqual(response.data["username"], str("foobarata"))  # type: ignore
+        self.assertEqual(response.data["email"], str("foobar@example.com"))  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:34]: follow DRY concept you are repeating response.data multiple  times

File: app/user/tests/test_views.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        link = reverse(
+            "email-verify",
+            kwargs={
+                "uidb64": urlsafe_base64_encode(force_bytes(new_user.pk)),
+                "token": account_activation_token.make_token(new_user),
+            },
+        )
+        resp = self.client.get(link)
+
+        self.assertEqual(resp.status_code, 200)
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:50]: assert more than once

File: app/user/tests/test_views.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        data = {
+            "username": "foobarata",
+            "email": "foobar@example.com",
+            "password": "somepassword",
+        }
+
+        response = self.client.post(self.create_url, data, format="json")
+
+        self.assertNotEqual(response, status.HTTP_201_CREATED)
+        self.assertEqual(response.data["username"], str("foobarata"))  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:33]: assert email is sent

File: app/user/token.py [hunk 22]
```diff
@@ -0,0 +1,16 @@
+from typing import Any
+
+import six
```
Comment (mainamuriuki274) [inline app/user/token.py:3]: use django's smart_bytes

File: app/user/token.py [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+
+
+class TokenGenerator(PasswordResetTokenGenerator):
+    def _make_hash_value(self, user, timestamp) -> Any:  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/token.py:8]: remove type ignore, io ni shortcut do the right thing

File: app/user/views.py [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+    serializer_class = VerifyEmailSerializer
+
+    def get(
+        self, request: Request, uidb64: str, token: str, **kwargs: str
+    ) -> Response:
+        data = {"uidb64": uidb64, "token": token}
+
+        serializer = self.get_serializer(data=data)
+        serializer.is_valid(raise_exception=True)
+        return Response("Email verified", status=status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline app/user/views.py:39]: have the response string as a json(optional)

File: speaksfer/settings/base.py (hunk 8 above)
Comment (mainamuriuki274) [reply speaksfer/settings/base.py:46]: did you fix it?

File: speaksfer/urls.py [hunk 25]
```diff
@@ -32,6 +32,7 @@
 
 urlpatterns = [
     path("admin/", admin.site.urls),
+    path("api/", include("app.urls")),
     path(
```
Comment (mainamuriuki274) [inline speaksfer/urls.py:36]: instead of swagger/ you can have docs/ it's more descriptive (optional)

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: speaksfer/settings/base.py (hunk 8 above)
Comment (dalienst) [reply speaksfer/settings/base.py:46]: Yes I did

File: app/user/serializers.py (hunk 11 above)
Comment (dalienst) [reply app/user/serializers.py:67]: It is only ignoring that line 67.

Comment (dalienst) [reply app/user/serializers.py:67]: MyPy was failing to pass even when I specified the type

File: app/user/serializers.py (hunk 12 above)
Comment (dalienst) [reply app/user/serializers.py:32]: The Profile Serializer was created in the first branch we worked on when creating the models

File: Pipfile (hunk 9 above)
Comment (dalienst) [reply Pipfile]: Done

File: app/user/serializers.py (hunk 10 above)
Comment (dalienst) [reply app/user/serializers.py:22]: Done

File: app/user/serializers.py (hunk 11 above)
Comment (dalienst) [reply app/user/serializers.py:80]: I cannot find the correct type but I have put the type annotations everywhere else

**Review by ChrisOnsando — COMMENTED** [review]
The changes have been made. The PR is ready for merging

File: app/user/serializers.py [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+    @staticmethod
+    def send_email(user: Any, request: Request) -> None:
+
+        current_site_info = get_current_site(request)
+        email_body = render_to_string(
+            "email_verification.html",
+            {
+                "user": user,
+                "domain": current_site_info.domain,
+                "uid": urlsafe_base64_encode(force_bytes(user.pk)),
```
Comment (mainamuriuki274) [inline app/user/serializers.py:64]: Use non sequential primary key

File: app/user/serializers.py [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+            email_body,
+            EMAIL_USER,
+            [user.email],
+            fail_silently=False,
+        )
+
+    def create(self, validated_data: Any) -> Any:
+        request = self.context.get("request")
+        user = User.objects.create_user(**validated_data)
+        self.send_email(user, request)  # type: ignore
```
Comment (mainamuriuki274) [inline app/user/serializers.py:80]: No need for ignoring mypy here, fix the error

File: app/user/serializers.py (hunk 5 above)
Comment (mainamuriuki274) [reply app/user/serializers.py:88]: You have not resolved this!

File: app/user/serializers.py (hunk 13 above)
Comment (mainamuriuki274) [reply app/user/serializers.py:118]: Why are you marking it as resolved and you have not made the changes !?

File: app/user/serializers.py (hunk 5 above)
Comment (dalienst) [reply app/user/serializers.py:88]: The serializer was made when we were working on the user model branch

File: app/user/views.py (hunk 7 above)
Comment (dalienst) [reply app/user/views.py]: Changed

File: app/user/serializers.py (hunk 27 above)
Comment (dalienst) [reply app/user/serializers.py:80]: This was one of the ones I tried to resolve but it kept throwing an error

File: Pipfile [hunk 28]
```diff
@@ -12,6 +12,8 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+djangorestframework-simplejwt = "5.2.0"
+faker = "13.14.0"
```
[the remark this answers is not in the document]
Comment (mainamuriuki274) [reply Pipfile:16]: remove it as a dependency

File: app/user/serializers.py (hunk 5 above)
Comment (mainamuriuki274) [reply app/user/serializers.py:87]: Cause you should not even have the Profile because that belongs to another card

File: app/user/tests/test_serializer.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class ImageUploadTest(TestCase):
+    def test_get_image(self) -> None:
+        user = User.objects.create(
+            username=fake.name(),
+            email=fake.email(),
+            password=fake.password(),
+        )
+        data = Profile.objects.create(
```
Comment (mainamuriuki274) [inline app/user/tests/test_serializer.py:18]: Doesn't Profile belong to another card??

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

**Comment by ChrisOnsando** [comment]
Dalienst , make the changes requested

**Comment by ChrisOnsando** [comment]
I also experience the mypy errors when  i do not ignore

File: app/user/serializers.py [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+            user = User.objects.get(pk=user_id)
+
+        except (TypeError, ValueError, OverflowError, User.DoesNotExist):
+            raise serializers.ValidationError(
+                "Invalid user id", code="invalid_code"
+            )
+
+        token = data.get("token")
+        if user and account_activation_token.check_token(user, token):
+
```
Comment (mainamuriuki274) [inline app/user/serializers.py:103]: remove this empty line

File: app/user/tests/test_views.py [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+from faker import Faker
+from rest_framework import status
+
+from app.user.token import account_activation_token
+
+User = get_user_model()
+fake = Faker()
+
+
+class ViewsTest(TestCase):
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:15]: Change the name to the View you are testing

File: app/user/tests/test_views.py [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+    """Testing views"""
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:16]: give a more decriptive docstring

File: app/user/token.py [hunk 33]
```diff
@@ -0,0 +1,16 @@
...
+
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+from django.utils.encoding import smart_bytes
+
+
+class TokenGenerator(PasswordResetTokenGenerator):
+    def _make_hash_value(self, user: Any, timestamp: Any) -> Any:
+        return (
+            smart_bytes(user.pk)
+            + smart_bytes(timestamp)
```
Comment (mainamuriuki274) [inline app/user/token.py:11]: prefer to use format string over string concatenation

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/user/token.py (hunk 33 above)
Comment (dalienst) [reply app/user/token.py:11]: I get an attribute error. Bytes object has no attribute format

**Review by mainamuriuki274 — APPROVED** [review]
